crates/pedersen/src/hash.rs

```diff
@@ -0,0 +1,534 @@
+    pub fn view_bits(&self) -> &BitSlice<Msb0, u8> {
+        &self.0.view_bits()[5..]
+    }
+
+    /// Creates a [StarkHash] from up-to 251 bits.
+    pub fn from_bits(bits: &BitSlice<Msb0, u8>) -> Result<Self, OverflowError> {
+        if bits.len() > 251 {
+            return Err(OverflowError);
+        }
+
+        let mut bytes = [0u8; 32];
+        bytes.view_bits_mut::<Msb0>()[256 - bits.len()..].copy_from_bitslice(bits);
+
+        Ok(Self(bytes))
+    }
+}
+
+impl std::ops::Add for StarkHash {
+    type Output = StarkHash;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        let result = FieldElement::from(self) + FieldElement::from(rhs);
+        StarkHash::from(result)
+    }
+}
+
+/// Computes the [Starknet Pedersen hash] on `a` and `b`.
+///
+/// [Starknet Pedersen hash]: https://docs.starkware.co/starkex-v3/crypto/pedersen-hash-function
+pub fn pedersen_hash(a: StarkHash, b: StarkHash) -> StarkHash {
+    let a = FieldElement::from(a).into_bits();
+    let b = FieldElement::from(b).into_bits();
+
+    // Compute hash
+    let mut accumulator = PEDERSEN_P0;
+    accumulator.add(&PEDERSEN_P1.multiply(&a[..248])); // Add a_low * P1
+    accumulator.add(&PEDERSEN_P2.multiply(&a[248..252])); // Add a_high * P2
+    accumulator.add(&PEDERSEN_P3.multiply(&b[..248])); // Add b_low * P3
+    accumulator.add(&PEDERSEN_P4.multiply(&b[248..252])); // Add b_high * P4
+
+    // Convert to affine
+    let result = AffinePoint::from(&accumulator);
+
+    // Return x-coordinate
+    StarkHash::from(result.x)
+}
+
+/// Computes the [Starknet Pedersen hash] on `a` and `b` using precomputed points.
+///
+/// [Starknet Pedersen hash]: https://docs.starkware.co/starkex-v3/crypto/pedersen-hash-function
+pub fn pedersen_hash_preprocessed(a: StarkHash, b: StarkHash) -> StarkHash {
+    let a = FieldElement::from(a).into_bits();
+    let b = FieldElement::from(b).into_bits();
+
+    // Preprocessed material is lookup-tables for each chunk of bits
+    let table_size = (1 << CURVE_CONSTS_BITS) - 1;
+    let add_points = |acc: &mut ProjectivePoint, bits: &BitSlice<_, u64>, prep: &[AffinePoint]| {
+        bits.chunks(CURVE_CONSTS_BITS)
+            .enumerate()
+            .for_each(|(i, v)| {
+                let offset: usize = v.load_le();
+                if offset > 0 {
+                    // Table lookup at 'offset-1' in table for chunk 'i'
+                    acc.add_affine(&prep[i * table_size + offset - 1]);
+                }
+            });
+    };
+
+    // Compute hash
+    let mut acc = PEDERSEN_P0;
+    add_points(&mut acc, &a[..248], &CURVE_CONSTS_P1); // Add a_low * P1
+    add_points(&mut acc, &a[248..252], &CURVE_CONSTS_P2); // Add a_high * P2
+    add_points(&mut acc, &b[..248], &CURVE_CONSTS_P3); // Add b_low * P3
+    add_points(&mut acc, &b[248..252], &CURVE_CONSTS_P4); // Add b_high * P4
+
+    // Convert to affine
+    let result = AffinePoint::from(&acc);
+
+    // Return x-coordinate
+    StarkHash::from(result.x)
+}
```

Comment:
I presume we would only keep the faster version of these two functions? Or is it something you'd like to keep to test against? If the latter I would move it to the test module.

Reply:
I think it's nice to have for completeness, since the crate is moving towards a crypto crate more than hash only, and shouldn't impact a final binary anyway. Otherwise we could shave off a lot of code with the same reasoning, for instance we only strictly need `projective::add_affine` and the `CURVE_CONSTS_*` for the preprocessed hash.

Reply:
> I would also consider updating the README to include instructions on how to generate the curve etc? Or maybe as docs on the function. (if that's applicable).

I added Joonas' proposals, so it is now referenced in `const_curve_gen.rs` and `curve_consts.rs`. Do you think this suffice? I think the main README is overkill, maybe we could have a separate one for the crypto/pedersen crate.

Reply:
Should be fine for now.